#include "MainWindow.h"
#include "ui_MainWindow.h"
#include "EditDialog.h"
#include "OperationsWindow.h"

#include <QFileDialog>
#include <QMessageBox>
#include <QFile>
#include <QTextStream>
#include <fstream>
#include <sstream>
#include <stdexcept>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , editDialog(nullptr)
    , operationsWindow(nullptr)
{
    ui->setupUi(this);
    setWindowTitle("–ú–µ–Ω–µ–¥–∂–µ—Ä –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤");

    // –ù–∞–ª–∞—à—Ç—É–≤–∞–Ω–Ω—è —Ç–∞–±–ª–∏—Ü—ñ
    ui->tableWidget->setColumnCount(4);
    QStringList headers = {"–ú–∞—Ä–∫–∞", "–ö–æ–ª—ñ—Ä", "–¶—ñ–Ω–∞ ($)", "–ü–æ—Ç—É–∂–Ω—ñ—Å—Ç—å (–∫.—Å.)"};
    ui->tableWidget->setHorizontalHeaderLabels(headers);
    ui->tableWidget->horizontalHeader()->setStretchLastSection(true);
    ui->tableWidget->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableWidget->setAlternatingRowColors(true);

    // –ù–∞–ª–∞—à—Ç—É–≤–∞–Ω–Ω—è –ø–æ—à—É–∫—É
    ui->searchText->setPlaceholderText("üîç –ü–æ—à—É–∫ –∑–∞ –º–∞—Ä–∫–æ—é –∞–±–æ –∫–æ–ª—å–æ—Ä–æ–º...");

    // –°—Ç–∏–ª—ñ–∑–∞—Ü—ñ—è –∫–Ω–æ–ø–æ–∫
    QString buttonStyle = "QPushButton { "
                          "padding: 8px 16px; "
                          "border: 1px solid #ccc; "
                          "border-radius: 4px; "
                          "background-color: #f0f0f0; "
                          "min-width: 80px; "
                          "}"
                          "QPushButton:hover { "
                          "background-color: #e0e0e0; "
                          "}"
                          "QPushButton:pressed { "
                          "background-color: #d0d0d0; "
                          "}";

    ui->addButton->setStyleSheet(buttonStyle);
    ui->editButton->setStyleSheet(buttonStyle);
    ui->deleteButton->setStyleSheet(buttonStyle);
    ui->loadButton->setStyleSheet(buttonStyle);
    ui->saveButton->setStyleSheet(buttonStyle);
    ui->operationsButton->setStyleSheet(buttonStyle);

    showStatusMessage("–ì–æ—Ç–æ–≤–æ –¥–æ —Ä–æ–±–æ—Ç–∏. –û–±–µ—Ä—ñ—Ç—å –¥—ñ—é –∑ –º–µ–Ω—é –≤–∏—â–µ.");
}

MainWindow::~MainWindow()
{
    delete ui;
    delete editDialog;
    delete operationsWindow;
}

void MainWindow::updateTable()
{
    applyFilter(ui->searchText->text());
}

void MainWindow::applyFilter(const QString& filter)
{
    if (filter.isEmpty()) {
        filteredCars = cars;
    } else {
        filteredCars.clear();
        QString lowerFilter = filter.toLower();
        for (const auto& car : cars) {
            if (QString::fromStdString(car.brand).toLower().contains(lowerFilter) ||
                QString::fromStdString(car.color).toLower().contains(lowerFilter)) {
                filteredCars.append(car);
            }
        }
    }

    ui->tableWidget->setRowCount(filteredCars.size());

    for (int i = 0; i < filteredCars.size(); ++i) {
        const Car& car = filteredCars[i];
        ui->tableWidget->setItem(i, 0, new QTableWidgetItem(QString::fromStdString(car.brand)));
        ui->tableWidget->setItem(i, 1, new QTableWidgetItem(QString::fromStdString(car.color)));
        ui->tableWidget->setItem(i, 2, new QTableWidgetItem(QString::number(car.price, 'f', 2)));
        ui->tableWidget->setItem(i, 3, new QTableWidgetItem(QString::number(car.power)));
    }

    // –û–Ω–æ–≤–ª–µ–Ω–Ω—è —Å—Ç–∞—Ç—É—Å—É
    if (filteredCars.isEmpty()) {
        if (cars.isEmpty()) {
            showStatusMessage("–¢–∞–±–ª–∏—Ü—è –ø–æ—Ä–æ–∂–Ω—è. –î–æ–¥–∞–π—Ç–µ –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ –∞–±–æ –∑–∞–≤–∞–Ω—Ç–∞–∂—Ç–µ –∑ —Ñ–∞–π–ª—É.");
        } else {
            showStatusMessage("–ù–µ –∑–Ω–∞–π–¥–µ–Ω–æ –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤ –∑–∞ –∑–∞–ø–∏—Ç–æ–º: " + filter);
        }
    } else {
        QString status = QString("–ü–æ–∫–∞–∑–∞–Ω–æ %1 –∑ %2 –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤").arg(filteredCars.size()).arg(cars.size());
        if (!filter.isEmpty()) {
            status += " (—Ñ—ñ–ª—å—Ç—Ä: " + filter + ")";
        }
        showStatusMessage(status);
    }
}

void MainWindow::showStatusMessage(const QString& message, bool isError)
{
    QString styledMessage = message;
    if (isError) {
        styledMessage = "‚ùå " + message;
    }
    ui->statusbar->showMessage(styledMessage);
}

bool MainWindow::isValidFileExtension(const QString& filePath) {
    QFileInfo fileInfo(filePath);
    QString extension = fileInfo.suffix().toLower();
    return (extension == "txt" || extension == "dat" || extension == "csv");
}

void MainWindow::validateFileForReading(const QString& filePath) {
    if (!isValidFileExtension(filePath)) {
        throw std::invalid_argument("–ù–µ–ø—ñ–¥—Ç—Ä–∏–º—É–≤–∞–Ω–∏–π —Ç–∏–ø —Ñ–∞–π–ª—É. –î–æ–∑–≤–æ–ª–µ–Ω—ñ —Ñ–æ—Ä–º–∞—Ç–∏: .txt, .dat, .csv");
    }

    QFile file(filePath);
    if (!file.exists()) {
        throw std::runtime_error("–§–∞–π–ª –Ω–µ —ñ—Å–Ω—É—î: " + filePath.toStdString());
    }

    if (file.size() == 0) {
        throw std::runtime_error("–§–∞–π–ª –ø–æ—Ä–æ–∂–Ω—ñ–π: " + filePath.toStdString());
    }
}

void MainWindow::validateFileForWriting(const QString& filePath) {
    if (!isValidFileExtension(filePath)) {
        throw std::invalid_argument("–ù–µ–ø—ñ–¥—Ç—Ä–∏–º—É–≤–∞–Ω–∏–π —Ç–∏–ø —Ñ–∞–π–ª—É. –î–æ–∑–≤–æ–ª–µ–Ω—ñ —Ñ–æ—Ä–º–∞—Ç–∏: .txt, .dat, .csv");
    }

    QFileInfo fileInfo(filePath);
    if (fileInfo.exists() && !fileInfo.isWritable()) {
        throw std::runtime_error("–ù–µ–º–∞—î –ø—Ä–∞–≤ –∑–∞–ø–∏—Å—É —É —Ñ–∞–π–ª: " + filePath.toStdString());
    }
}

void MainWindow::loadFromFile(const QString& filePath) {
    try {
        validateFileForReading(filePath);

        std::ifstream file(filePath.toStdString());
        if (!file.is_open()) {
            throw std::runtime_error("–ù–µ –≤–¥–∞–ª–æ—Å—è –≤—ñ–¥–∫—Ä–∏—Ç–∏ —Ñ–∞–π–ª –¥–ª—è —á–∏—Ç–∞–Ω–Ω—è");
        }

        cars.clear();
        Car car;
        int loadedCount = 0;
        int lineNumber = 0;
        std::string line;

        while (std::getline(file, line)) {
            lineNumber++;
            if (line.empty()) continue;

            try {
                std::stringstream ss(line);
                if (ss >> car) {
                    cars.append(car);
                    loadedCount++;
                }
            } catch (const std::exception& e) {
                showStatusMessage(QString("–ü–æ–º–∏–ª–∫–∞ —É —Ä—è–¥–∫—É %1: %2").arg(lineNumber).arg(e.what()), true);
            }
        }
        file.close();

        ui->searchText->clear();
        updateTable();

        if (loadedCount > 0) {
            showStatusMessage("–£—Å–ø—ñ—à–Ω–æ –∑–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–æ " + QString::number(loadedCount) + " –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤ –∑ —Ñ–∞–π–ª—É");
            QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                     "–£—Å–ø—ñ—à–Ω–æ –∑–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–æ " + QString::number(loadedCount) + " –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤ –∑ —Ñ–∞–π–ª—É.");
        } else {
            throw std::runtime_error("–ù–µ –∑–Ω–∞–π–¥–µ–Ω–æ –∂–æ–¥–Ω–æ–≥–æ –∫–æ—Ä–µ–∫—Ç–Ω–æ–≥–æ –∑–∞–ø–∏—Å—É —É —Ñ–∞–π–ª—ñ");
        }
    } catch (const std::exception& e) {
        showStatusMessage("–ü–æ–º–∏–ª–∫–∞ –∑–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–Ω—è: " + QString(e.what()), true);
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ –∑–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–Ω—è",
                              QString("–ù–µ –≤–¥–∞–ª–æ—Å—è –∑–∞–≤–∞–Ω—Ç–∞–∂–∏—Ç–∏ —Ñ–∞–π–ª:\n%1").arg(e.what()));
    }
}

void MainWindow::on_addButton_clicked() {
    try {
        if (!editDialog) {
            editDialog = new EditDialog(this);
        }

        editDialog->setCarData(Car());
        if (editDialog->exec() == QDialog::Accepted) {
            Car newCar = editDialog->getCarData();
            cars.append(newCar);
            updateTable();

            showStatusMessage("–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –¥–æ–¥–∞–Ω–æ: " + QString::fromStdString(newCar.brand));
            QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                     QString("–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –¥–æ–¥–∞–Ω–æ!\n%1 - %2 - %3 - %4")
                                         .arg(QString::fromStdString(newCar.brand))
                                         .arg(QString::fromStdString(newCar.color))
                                         .arg(newCar.price)
                                         .arg(newCar.power));
        }
    } catch (const std::exception& e) {
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ –¥–æ–¥–∞–≤–∞–Ω–Ω—è", QString("–ü–æ–º–∏–ª–∫–∞ –ø—Ä–∏ –¥–æ–¥–∞–≤–∞–Ω–Ω—ñ –∞–≤—Ç–æ: %1").arg(e.what()));
    }
}

void MainWindow::on_editButton_clicked() {
    try {
        int currentRow = ui->tableWidget->currentRow();
        if (currentRow < 0 || currentRow >= filteredCars.size()) {
            throw std::runtime_error("–ë—É–¥—å –ª–∞—Å–∫–∞, –≤–∏–±–µ—Ä—ñ—Ç—å –∞–≤—Ç–æ–º–æ–±—ñ–ª—å –¥–ª—è —Ä–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è");
        }

        // –ó–Ω–∞—Ö–æ–¥–∏–º–æ –æ—Ä–∏–≥—ñ–Ω–∞–ª—å–Ω–∏–π —ñ–Ω–¥–µ–∫—Å —É –≥–æ–ª–æ–≤–Ω–æ–º—É –º–∞—Å–∏–≤—ñ
        Car selectedCar = filteredCars[currentRow];
        int originalIndex = -1;
        for (int i = 0; i < cars.size(); ++i) {
            if (cars[i] == selectedCar) {
                originalIndex = i;
                break;
            }
        }

        if (originalIndex == -1) {
            throw std::runtime_error("–ù–µ –≤–¥–∞–ª–æ—Å—è –∑–Ω–∞–π—Ç–∏ –æ–±—Ä–∞–Ω–∏–π –∞–≤—Ç–æ–º–æ–±—ñ–ª—å");
        }

        if (!editDialog) {
            editDialog = new EditDialog(this);
        }

        Car originalCar = cars[originalIndex];
        editDialog->setCarData(originalCar);
        if (editDialog->exec() == QDialog::Accepted) {
            Car updatedCar = editDialog->getCarData();
            cars[originalIndex] = updatedCar;
            updateTable();

            showStatusMessage("–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –æ–Ω–æ–≤–ª–µ–Ω–æ: " + QString::fromStdString(updatedCar.brand));
            QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                     QString("–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –æ–Ω–æ–≤–ª–µ–Ω–æ!\n–ë—É–ª–æ: %1 - %2 - %3 - %4\n–°—Ç–∞–ª–æ: %5 - %6 - %7 - %8")
                                         .arg(QString::fromStdString(originalCar.brand))
                                         .arg(QString::fromStdString(originalCar.color))
                                         .arg(originalCar.price)
                                         .arg(originalCar.power)
                                         .arg(QString::fromStdString(updatedCar.brand))
                                         .arg(QString::fromStdString(updatedCar.color))
                                         .arg(updatedCar.price)
                                         .arg(updatedCar.power));
        }
    } catch (const std::exception& e) {
        showStatusMessage("–ü–æ–º–∏–ª–∫–∞ —Ä–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è: " + QString(e.what()), true);
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ —Ä–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è", e.what());
    }
}

void MainWindow::on_deleteButton_clicked() {
    try {
        int currentRow = ui->tableWidget->currentRow();
        if (currentRow < 0 || currentRow >= filteredCars.size()) {
            throw std::runtime_error("–ë—É–¥—å –ª–∞—Å–∫–∞, –≤–∏–±–µ—Ä—ñ—Ç—å –∞–≤—Ç–æ–º–æ–±—ñ–ª—å –¥–ª—è –≤–∏–¥–∞–ª–µ–Ω–Ω—è");
        }

        Car carToDelete = filteredCars[currentRow];
        int originalIndex = -1;
        for (int i = 0; i < cars.size(); ++i) {
            if (cars[i] == carToDelete) {
                originalIndex = i;
                break;
            }
        }

        if (originalIndex == -1) {
            throw std::runtime_error("–ù–µ –≤–¥–∞–ª–æ—Å—è –∑–Ω–∞–π—Ç–∏ –æ–±—Ä–∞–Ω–∏–π –∞–≤—Ç–æ–º–æ–±—ñ–ª—å");
        }

        QMessageBox::StandardButton reply = QMessageBox::question(this, "–ü—ñ–¥—Ç–≤–µ—Ä–¥–∂–µ–Ω–Ω—è –≤–∏–¥–∞–ª–µ–Ω–Ω—è",
                                                                  QString("–í–∏ –≤–ø–µ–≤–Ω–µ–Ω—ñ, —â–æ —Ö–æ—á–µ—Ç–µ –≤–∏–¥–∞–ª–∏—Ç–∏ —Ü–µ–π –∞–≤—Ç–æ–º–æ–±—ñ–ª—å?\n%1 - %2 - %3 - %4")
                                                                      .arg(QString::fromStdString(carToDelete.brand))
                                                                      .arg(QString::fromStdString(carToDelete.color))
                                                                      .arg(carToDelete.price)
                                                                      .arg(carToDelete.power),
                                                                  QMessageBox::Yes | QMessageBox::No);

        if (reply == QMessageBox::Yes) {
            cars.removeAt(originalIndex);
            updateTable();
            showStatusMessage("–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –≤–∏–¥–∞–ª–µ–Ω–æ");
            QMessageBox::information(this, "–£—Å–ø—ñ—Ö", "–ê–≤—Ç–æ–º–æ–±—ñ–ª—å —É—Å–ø—ñ—à–Ω–æ –≤–∏–¥–∞–ª–µ–Ω–æ.");
        }
    } catch (const std::exception& e) {
        showStatusMessage("–ü–æ–º–∏–ª–∫–∞ –≤–∏–¥–∞–ª–µ–Ω–Ω—è: " + QString(e.what()), true);
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ –≤–∏–¥–∞–ª–µ–Ω–Ω—è", e.what());
    }
}

void MainWindow::on_loadButton_clicked() {
    try {
        QString filePath = QFileDialog::getOpenFileName(this,
                                                        "–í–∏–±–µ—Ä—ñ—Ç—å —Ñ–∞–π–ª –¥–ª—è –∑–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–Ω—è",
                                                        "",
                                                        "–¢–µ–∫—Å—Ç–æ–≤—ñ —Ñ–∞–π–ª–∏ (*.txt);;–§–∞–π–ª–∏ –¥–∞–Ω–∏—Ö (*.dat);;CSV —Ñ–∞–π–ª–∏ (*.csv);;–í—Å—ñ —Ñ–∞–π–ª–∏ (*)");

        if (!filePath.isEmpty()) {
            loadFromFile(filePath);
        } else {
            showStatusMessage("–ó–∞–≤–∞–Ω—Ç–∞–∂–µ–Ω–Ω—è —Å–∫–∞—Å–æ–≤–∞–Ω–æ");
        }
    } catch (const std::exception& e) {
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞", QString("–ü–æ–º–∏–ª–∫–∞ –≤–∏–±–æ—Ä—É —Ñ–∞–π–ª—É: %1").arg(e.what()));
    }
}

void MainWindow::on_saveButton_clicked() {
    try {
        if (cars.isEmpty()) {
            throw std::runtime_error("–ù–µ–º–∞—î –¥–∞–Ω–∏—Ö –¥–ª—è –∑–±–µ—Ä–µ–∂–µ–Ω–Ω—è");
        }

        QString filePath = QFileDialog::getSaveFileName(this,
                                                        "–ó–±–µ—Ä–µ–≥—Ç–∏ –¥–∞–Ω—ñ",
                                                        "",
                                                        "–¢–µ–∫—Å—Ç–æ–≤—ñ —Ñ–∞–π–ª–∏ (*.txt);;–§–∞–π–ª–∏ –¥–∞–Ω–∏—Ö (*.dat);;CSV —Ñ–∞–π–ª–∏ (*.csv);;–í—Å—ñ —Ñ–∞–π–ª–∏ (*)");

        if (!filePath.isEmpty()) {
            validateFileForWriting(filePath);

            std::ofstream file(filePath.toStdString());
            if (!file.is_open()) {
                throw std::runtime_error("–ù–µ –≤–¥–∞–ª–æ—Å—è —Å—Ç–≤–æ—Ä–∏—Ç–∏ —Ñ–∞–π–ª –¥–ª—è –∑–∞–ø–∏—Å—É");
            }

            int savedCount = 0;
            for (const auto& car : cars) {
                try {
                    file << car;
                    savedCount++;
                } catch (const std::exception& e) {
                    showStatusMessage("–ü–æ–º–∏–ª–∫–∞ –∑–±–µ—Ä–µ–∂–µ–Ω–Ω—è –∞–≤—Ç–æ–º–æ–±—ñ–ª—è: " + QString(e.what()), true);
                }
            }
            file.close();

            if (savedCount > 0) {
                showStatusMessage("–£—Å–ø—ñ—à–Ω–æ –∑–±–µ—Ä–µ–∂–µ–Ω–æ " + QString::number(savedCount) + " –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤ —É —Ñ–∞–π–ª");
                QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                         "–î–∞–Ω—ñ —É—Å–ø—ñ—à–Ω–æ –∑–±–µ—Ä–µ–∂–µ–Ω–æ!\n" +
                                             QString::number(savedCount) + " –∞–≤—Ç–æ–º–æ–±—ñ–ª—ñ–≤ –∑–±–µ—Ä–µ–∂–µ–Ω–æ —É —Ñ–∞–π–ª: " + filePath);
            } else {
                throw std::runtime_error("–ù–µ –≤–¥–∞–ª–æ—Å—è –∑–±–µ—Ä–µ–≥—Ç–∏ –∂–æ–¥–Ω–æ–≥–æ –∞–≤—Ç–æ–º–æ–±—ñ–ª—è");
            }
        } else {
            showStatusMessage("–ó–±–µ—Ä–µ–∂–µ–Ω–Ω—è —Å–∫–∞—Å–æ–≤–∞–Ω–æ");
        }
    } catch (const std::exception& e) {
        showStatusMessage("–ü–æ–º–∏–ª–∫–∞ –∑–±–µ—Ä–µ–∂–µ–Ω–Ω—è: " + QString(e.what()), true);
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ –∑–±–µ—Ä–µ–∂–µ–Ω–Ω—è",
                              QString("–ù–µ –≤–¥–∞–ª–æ—Å—è –∑–±–µ—Ä–µ–≥—Ç–∏ —Ñ–∞–π–ª:\n%1").arg(e.what()));
    }
}

void MainWindow::on_operationsButton_clicked() {
    try {
        if (cars.isEmpty()) {
            throw std::runtime_error("–ù–µ–º–∞—î –¥–∞–Ω–∏—Ö –¥–ª—è –æ–ø–µ—Ä–∞—Ü—ñ–π");
        }

        if (!operationsWindow) {
            operationsWindow = new OperationsWindow(this);
            connect(operationsWindow, &OperationsWindow::carsSorted, this, [this](const QVector<Car>& sortedCars) {
                this->cars = sortedCars;
                updateTable();
                showStatusMessage("–¢–∞–±–ª–∏—Ü—é —É—Å–ø—ñ—à–Ω–æ –≤—ñ–¥—Å–æ—Ä—Ç–æ–≤–∞–Ω–æ");
            });
        }

        operationsWindow->setCars(cars);
        operationsWindow->exec();
    } catch (const std::exception& e) {
        showStatusMessage("–ü–æ–º–∏–ª–∫–∞ –æ–ø–µ—Ä–∞—Ü—ñ–π: " + QString(e.what()), true);
        QMessageBox::critical(this, "–ü–æ–º–∏–ª–∫–∞ –æ–ø–µ—Ä–∞—Ü—ñ–π", e.what());
    }
}

void MainWindow::on_tableWidget_itemDoubleClicked(QTableWidgetItem *item) {
    Q_UNUSED(item)
    on_editButton_clicked();
}

void MainWindow::on_searchText_textChanged(const QString &text) {
    applyFilter(text);
}
